test/e2e/pod-scaler/prometheus/backfill.go

```diff
@@ -0,0 +1,378 @@
+// +build e2e
+
+package prometheus
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math"
+	"math/rand"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/openhistogram/circonusllhist"
+	prometheus_client "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/prometheus/pkg/value"
+	uuid "github.com/satori/go.uuid"
+
+	"k8s.io/utils/pointer"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	pod_scaler "github.com/openshift/ci-tools/pkg/pod-scaler"
+	"github.com/openshift/ci-tools/pkg/testhelper"
+)
+
+const (
+	// numSeries is the number of individual series of data generated for a given label set
+	numSeries = 10
+	// numSamples is the number of samples in an individual series
+	numSamples = 250
+	// samplingDuration is the duration between samples in a series
+	samplingDuration = time.Minute
+	// seriesDuration
+	seriesDuration = time.Duration(numSamples) * samplingDuration
+)
+
+type DataInStages struct {
+	ByOffset map[time.Duration]Data
+}
+
+type Data struct {
+	ByFile map[string]map[pod_scaler.FullMetadata][]*circonusllhist.HistogramWithoutLookups
+}
+
+func (d *DataInStages) record(offset time.Duration, metric string, labels seriesLabels, identifier pod_scaler.FullMetadata, data []float64) error {
+	hist := circonusllhist.New(circonusllhist.NoLookup(), circonusllhist.Size(1))
+	for _, v := range data {
+		if err := hist.RecordValue(v); err != nil {
+			return fmt.Errorf("could not insert value into histogram: %w", err)
+		}
+	}
+
+	if _, ok := d.ByOffset[offset]; !ok {
+		d.ByOffset[offset] = Data{
+			ByFile: map[string]map[pod_scaler.FullMetadata][]*circonusllhist.HistogramWithoutLookups{},
+		}
+	}
+	var file string
+	switch {
+	case labels["label_ci_openshift_io_metadata_step"] != "":
+		file = fmt.Sprintf("steps/%s.json", metric)
+	case labels["label_created_by_prow"] != "":
+		file = fmt.Sprintf("prowjobs/%s.json", metric)
+	default:
+		file = fmt.Sprintf("pods/%s.json", metric)
+	}
+	if _, ok := d.ByOffset[offset].ByFile[file]; !ok {
+		d.ByOffset[offset].ByFile[file] = map[pod_scaler.FullMetadata][]*circonusllhist.HistogramWithoutLookups{}
+	}
+	d.ByOffset[offset].ByFile[file][identifier] = append(d.ByOffset[offset].ByFile[file][identifier], circonusllhist.NewHistogramWithoutLookups(hist))
+
+	return nil
+}
+
+// Backfill generates data and backfills it into Prometheus, returning the data written by stage.
+func Backfill(t testhelper.TestingTInterface, prometheusDir string, retentionPeriod time.Duration) *DataInStages {
+	rand.Seed(time.Now().UnixNano())
```

Comment:
Would it be useful to write out the seed / allow to inject it, to allow reproducible runs?

Reply:
Yeah, good point